### PR TITLE
Cats blaze-client

### LINKED ISSUE #864: Port blaze-client to fs2-cats
See also #863.

@ChristopherDavenport has been working on this.

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
This is an initial blaze-client implementation. The client test battery had a severe error that caused some difficulties, but that is now taken care of. 

It has a single test which is commented out due to the Http Protocol being elevated to Http1.1 from 1.0

Closes #864

### CODE REVIEW

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeHttp1ClientParser.scala [hunk 1]
```diff
@@ -63,11 +63,11 @@ private final class BlazeHttp1ClientParser(maxResponseLineSize: Int,
                                             scheme: String,
                                             majorversion: Int,
                                             minorversion: Int): Unit = {
-    status = Status.fromIntAndReason(code, reason).valueOr(throw _)
+    status = Status.fromIntAndReason(code, reason).fold(fail => throw fail, status => status )
```
Comment (mkulak) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeHttp1ClientParser.scala]: `import cats.syntax.either._` will add exactly same `valueOr` method.

File: blaze-client/src/main/scala/org/http4s/client/blaze/BlazeHttp1ClientParser.scala [hunk 2]
```diff
@@ -63,11 +63,11 @@ private final class BlazeHttp1ClientParser(maxResponseLineSize: Int,
...
     httpVersion = {
       if (majorversion == 1 && minorversion == 1)  HttpVersion.`HTTP/1.1`
       else if (majorversion == 1 && minorversion == 0)  HttpVersion.`HTTP/1.0`
-      else HttpVersion.fromVersion(majorversion, minorversion).getOrElse(HttpVersion.`HTTP/1.0`)
+      else HttpVersion.fromVersion(majorversion, minorversion).fold(_ => HttpVersion.`HTTP/1.0`, httpVersion => httpVersion)
```
Comment (mkulak) [inline blaze-client/src/main/scala/org/http4s/client/blaze/BlazeHttp1ClientParser.scala]: Same here: cats can pimp `Either`

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 3]
```diff
@@ -134,18 +132,20 @@ private final class Http1Connection(val requestKey: RequestKey,
...
+          .writeEntityBody(req.body)
+          .handle { case EOF => false }
+        // If we get a pipeline closed, we might still be good. Check response
         val respTask = receiveResponse(mustClose, doesntHaveBody = req.method == Method.HEAD)
-        Task.taskInstance.mapBoth(bodyTask, respTask)((_,r) => r)
+        bodyTask.flatMap { _ => respTask
           .handleWith { case t =>
             fatalError(t, "Error executing request")
             Task.fail(t)
           }
```
Comment (mkulak) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: Perhaps consider `bodyTask.followedBy(respTask).handleWith { case t => ...`

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Support.scala [hunk 4]
```diff
@@ -77,12 +76,12 @@ final private class Http1Support(config: BlazeClientConfig, executor: ExecutorSe
...
   }
 
-  private def getAddress(requestKey: RequestKey): Throwable \/ InetSocketAddress = {
+  private def getAddress(requestKey: RequestKey): Either[Throwable, InetSocketAddress] = {
     requestKey match {
       case RequestKey(s, auth) =>
         val port = auth.port getOrElse { if (s == Https) 443 else 80 }
         val host = auth.host.value
-        \/.fromTryCatchNonFatal(new InetSocketAddress(host, port))
+        Either.fromTry(Try(new InetSocketAddress(host, port)))
```
Comment (mkulak) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Support.scala]: Consider `Either.catchNonFatal` from `cats.syntax.either._`

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 5]
```diff
@@ -179,11 +179,11 @@ private final class Http1Connection(val requestKey: RequestKey,
...
         // we are now to the body
-        def terminationCondition() = stageState.get match {  // if we don't have a length, EOF signals the end of the body.
-          case Error(e) if e != EOF => e
+        def terminationCondition(): Either[Throwable, Option[Chunk[Byte]]] = stageState.get match {  // if we don't have a length, EOF signals the end of the body.
+          case Error(e) if e != EOF => Left(e)
           case _ =>
-            if (parser.definedContentLength() || parser.isChunked()) InvalidBodyException("Received premature EOF.")
-            else Terminated(End)
+            if (parser.definedContentLength() || parser.isChunked()) Left(InvalidBodyException("Received premature EOF."))
+            else Right(None)
```
Comment (mkulak) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: Using `Either.right(None)` and `Either.left(InvalidBodyException ...)` will allow you to get rid of explicit return type annotation (after importing `cats.syntax.either._`).

**Comment by ChristopherDavenport** [comment]
@mkulak You are awesome! I was disappointed I didn't have a closer approximation, turns out I needed to get more comfortable with cats. I'll incorporate asap.

**Comment by ChristopherDavenport** [comment]
We should either extend our method length limit, or do some internal refactoring here some point in the future. That scalastyle error is really annoying as so many of the methods in this section are much longer than 50 lines.

File: blaze-client/src/test/scala/org/http4s/client/blaze/Http1ClientStageSpec.scala [hunk 6]
```diff
@@ -222,16 +220,17 @@ class Http1ClientStageSpec extends Http4sSpec {
...
+//    "Allow an HTTP/1.0 request without a Host header" in {
+//      val resp = "HTTP/1.0 200 OK\r\n\r\ndone"
+//
+//      val req = Request(uri = www_foo_test, httpVersion = HttpVersion.`HTTP/1.0`)
+//
+//      val (request, response) = getSubmission(req, resp)
+//
+//      request must not contain("Host:")
+//      response must_==("done")
+//    }
```
Comment (rossabaker) [inline blaze-client/src/test/scala/org/http4s/client/blaze/Http1ClientStageSpec.scala]: `pendingUntilFixed` keeps this compiling and notifies us if we accidentally fix it.

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala [hunk 7]
```diff
@@ -197,44 +204,38 @@ private final class Http1Connection(val requestKey: RequestKey,
...
           // We are to the point of parsing the body and then cleaning up
-          val (rawBody, _) = collectBodyFromParser(buffer, terminationCondition _)
+          val (rawBody, _): (EntityBody, () => Future[ByteBuffer] ) = collectBodyFromParser(buffer, terminationCondition _)
 
           // to collect the trailers we need a cleanup helper and a Task in the attribute map
-          val (trailerCleanup, attributes) =
+          val (trailerCleanup, attributes) : (()=> Unit, AttributeMap) = {
             if (parser.getHttpVersion().minor == 1 && parser.isChunked()) {
-              val trailers = new AtomicReference(Headers.empty)
+              val trailers: AtomicReference[Headers] = new AtomicReference(Headers.empty)
```
Comment (rossabaker) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: Is Scalastyle demanding these annotations?  I like it for public values, but it seems noisy for inner calculations.

File: blaze-client/src/test/scala/org/http4s/client/blaze/ClientTimeoutSpec.scala [hunk 8]
```diff
@@ -42,44 +41,44 @@ class ClientTimeoutSpec extends Http4sSpec {
...
 
-      c.fetchAs[String](FooRequest).run must throwA[TimeoutException]
+      c.fetchAs[String](FooRequest).unsafeRun() must throwA[TimeoutException]
     }
 
     "Request timeout on slow POST body" in {
 
       def dataStream(n: Int): EntityBody = {
-        implicit def defaultSecheduler = DefaultTimeoutScheduler
+        implicit val defaultScheduler = Scheduler.fromFixedDaemonPool(2)
```
Comment (rossabaker) [inline blaze-client/src/test/scala/org/http4s/client/blaze/ClientTimeoutSpec.scala]: I think this is going to leak thread pools in our tests.  Where is this being shut down?

There's also an `Http4s.TestPool` defined.

File: blaze-client/src/test/scala/org/http4s/client/blaze/Http1ClientStageSpec.scala [hunk 9]
```diff
@@ -13,14 +13,13 @@ import scodec.bits.ByteVector
...
-import scalaz.\/-
-import scalaz.concurrent.{Strategy, Task}
+import fs2._
 
 // TODO: this needs more tests
 class Http1ClientStageSpec extends Http4sSpec {
 
   val ec = org.http4s.blaze.util.Execution.trampoline
-  val es = Strategy.DefaultExecutorService
+  val es = impl.DefaultExecutor.newClientDefaultExecutorService("Http1ClientStageSpec")
```
Comment (rossabaker) [inline blaze-client/src/test/scala/org/http4s/client/blaze/Http1ClientStageSpec.scala]: Same problem as above.  We need to shut down the non-defaults or reuse the `Http4sSpec` one.

File: blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala (hunk 7 above)
Comment (ChristopherDavenport) [inline blaze-client/src/main/scala/org/http4s/client/blaze/Http1Connection.scala]: It's not, I was just finding it difficult to reason at time. The one you noted is absolutely redundant, but quite a few I had to do some reverse engineering to figure out.

**Pull request merged**